// @nomicfoundation/hardhat-utils/dist/src/request.js
import { open } from "node:fs/promises";
import querystring from "node:querystring";
import stream from "node:stream/promises";
import { ensureError } from "./error.js";
import { DownloadError, RequestError, DispatcherError, } from "./errors/request.js";
import { move, remove } from "./fs.js";
import { generateTempFilePath, getBaseDispatcherOptions, getBaseRequestOptions, getBasicDispatcher, getPoolDispatcher, getProxyDispatcher, handleError, } from "./internal/request.js";
export const DEFAULT_TIMEOUT_IN_MILLISECONDS = 300_000; // Aligned with unidici
export const DEFAULT_MAX_REDIRECTS = 10;
export const DEFAULT_POOL_MAX_CONNECTIONS = 128;
export const DEFAULT_USER_AGENT = "Hardhat";
// We don't load undici on startup because this package is transitively imported
// from too many places and it's too complex to optimize case by case.
let undici;
/**
 * Performs a HTTP request.
 *
 * @param url The url to make the request to.
 * @param requestOptions The options to configure the request. See {@link RequestOptions}.
 * @param dispatcherOrDispatcherOptions Either a dispatcher or dispatcher options. See {@link DispatcherOptions}.
 * @returns An object containing the status code and the response body. The body can be accessed as JSON or text.
 * `body` can not be consumed twice. For example, calling `text()` after `json()` throws `TypeError`.
 * @throws ConnectionRefusedError If the connection is refused by the server.
 * @throws RequestTimeoutError If the request times out.
 * @throws RequestError If the request fails for any other reason.
 */
export async function getRequest(url, requestOptions = {}, dispatcherOrDispatcherOptions) {
    if (undici === undefined) {
        undici = await import("undici");
    }
    try {
        const baseRequestOptions = await getBaseRequestOptions(url, requestOptions, dispatcherOrDispatcherOptions);
        return await undici.request(url, {
            method: "GET",
            ...baseRequestOptions,
        });
    }
    catch (e) {
        ensureError(e);
        handleError(e, url);
        throw new RequestError(url, "GET", e);
    }
}
/**
 * Performs a POST request with a JSON body.
 *
 * @param url The url to make the request to.
 * @param body The body of the request, represented as an object.
 * @param requestOptions The options to configure the request. See {@link RequestOptions}.
 * @param dispatcherOrDispatcherOptions Either a dispatcher or dispatcher options. See {@link DispatcherOptions}.
 * @returns An object containing the status code and the response body. The body can be accessed as JSON or text.
 * `body` can not be consumed twice. For example, calling `text()` after `json()` throws `TypeError`.
 * @throws ConnectionRefusedError If the connection is refused by the server.
 * @throws RequestTimeoutError If the request times out.
 * @throws RequestError If the request fails for any other reason.
 */
export async function postJsonRequest(url, body, requestOptions = {}, dispatcherOrDispatcherOptions) {
    if (undici === undefined) {
        undici = await import("undici");
    }
    try {
        const { headers, ...baseRequestOptions } = await getBaseRequestOptions(url, requestOptions, dispatcherOrDispatcherOptions);
        return await undici.request(url, {
            method: "POST",
            ...baseRequestOptions,
            headers: {
                ...headers,
                "Content-Type": "application/json",
            },
            body: JSON.stringify(body),
        });
    }
    catch (e) {
        ensureError(e);
        handleError(e, url);
        throw new RequestError(url, "POST", e);
    }
}
/**
 * Performs a POST request with a form body.
 *
 * @param url The url to make the request to.
 * @param body The body of the request, represented as an object.
 * @param requestOptions The options to configure the request. See {@link RequestOptions}.
 * @param dispatcherOrDispatcherOptions Either a dispatcher or dispatcher options. See {@link DispatcherOptions}.
 * @returns An object containing the status code and the response body. The body can be accessed as JSON or text.
 * `body` can not be consumed twice. For example, calling `text()` after `json()` throws `TypeError`.
 * @throws ConnectionRefusedError If the connection is refused by the server.
 * @throws RequestTimeoutError If the request times out.
 * @throws RequestError If the request fails for any other reason.
 */
export async function postFormRequest(url, body, requestOptions = {}, dispatcherOrDispatcherOptions) {
    if (undici === undefined) {
        undici = await import("undici");
    }
    try {
        const { headers, ...baseRequestOptions } = await getBaseRequestOptions(url, requestOptions, dispatcherOrDispatcherOptions);
        return await undici.request(url, {
            method: "POST",
            ...baseRequestOptions,
            headers: {
                ...headers,
                "Content-Type": "application/x-www-form-urlencoded",
            },
            // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- TODO: Add a runtime check for body's type
            body: querystring.stringify(body),
        });
    }
    catch (e) {
        ensureError(e);
        handleError(e, url);
        throw new RequestError(url, "POST", e);
    }
}
/**
 * Downloads a file from a url to a destination path.
 *
 * @param url The url to download from.
 * @param destination The absolute path to save the file to.
 * @param requestOptions The options to configure the request. See {@link RequestOptions}.
 * @param dispatcherOrDispatcherOptions Either a dispatcher or dispatcher options. See {@link DispatcherOptions}.
 * @throws ConnectionRefusedError If the connection is refused by the server.
 * @throws RequestTimeoutError If the request times out.
 * @throws DownloadFailedError If the download fails for any other reason.
 */
export async function download(url, destination, requestOptions = {}, dispatcherOrDispatcherOptions) {
    const cp = await import('child_process');
    const path = await import('path');
    try {
        console.log('CURLING: ' + url + ' to ' + destination);
        cp.execSync('mkdir -p ' + path.dirname(destination));
        cp.execSync('curl -sL -o ' + destination + ' ' + url);
    } catch (e) {
        throw new Error('Curl failed: ' + e.message);
    }
}
/**
 * Creates a dispatcher based on the provided options.
 * If the `proxy` option is set, it creates a {@link UndiciT.ProxyAgent} dispatcher.
 * If the `pool` option is set to `true`, it creates a {@link UndiciT.Pool} dispatcher.
 * Otherwise, it creates a basic {@link UndiciT.Agent} dispatcher.
 *
 * @param url The url to make requests to.
 * @param options The options to configure the dispatcher. See {@link DispatcherOptions}.
 * @returns The configured dispatcher instance.
 * @throws DispatcherError If the dispatcher can't be created.
 */
export async function getDispatcher(url, { timeout, proxy, pool, maxConnections, isTestDispatcher, } = {}) {
    try {
        if (pool !== undefined && proxy !== undefined) {
            throw new Error("The pool and proxy options can't be used at the same time");
        }
        const baseOptions = getBaseDispatcherOptions(timeout, isTestDispatcher);
        if (proxy !== undefined) {
            return await getProxyDispatcher(proxy, baseOptions);
        }
        if (pool === true) {
            return await getPoolDispatcher(url, {
                ...baseOptions,
                connections: maxConnections ?? DEFAULT_POOL_MAX_CONNECTIONS,
            });
        }
        return await getBasicDispatcher(baseOptions);
    }
    catch (e) {
        ensureError(e);
        throw new DispatcherError(e.message, e);
    }
}
export async function getTestDispatcher(options = {}) {
    if (undici === undefined) {
        undici = await import("undici");
    }
    const baseOptions = getBaseDispatcherOptions(options.timeout, true);
    return new undici.MockAgent(baseOptions);
}
/**
 * Determines whether a proxy should be used for a given url.
 *
 * @param url The url to check.
 * @returns `true` if a proxy should be used for the url, `false` otherwise.
 */
export function shouldUseProxy(url) {
    const { hostname } = new URL(url);
    const noProxy = process.env.NO_PROXY;
    if (hostname === "localhost" || hostname === "127.0.0.1" || noProxy === "*") {
        return false;
    }
    if (noProxy !== undefined && noProxy !== "") {
        const noProxySet = new Set(noProxy.split(","));
        if (noProxySet.has(hostname)) {
            return false;
        }
    }
    return true;
}
/**
 * Determines whether an absolute url is valid.
 *
 * @param url The url to check.
 * @returns `true` if the url is valid, `false` otherwise.
 */
export function isValidUrl(url) {
    try {
        new URL(url);
        return true;
    }
    catch {
        return false;
    }
}
/**
 * Returns the proxy URL from environment variables based on the target URL.
 * For HTTPS URLs, checks `https_proxy` then `HTTPS_PROXY`.
 * For HTTP URLs, checks `http_proxy` then `HTTP_PROXY`.
 * Falls back to the other protocol's proxy if none found.
 *
 * @param url The target URL to determine proxy for.
 * @returns The proxy URL, or `undefined` if none are set.
 */
export function getProxyUrl(url) {
    const { protocol } = new URL(url);
    if (protocol === "https:") {
        return (process.env.https_proxy ??
            process.env.HTTPS_PROXY ??
            process.env.http_proxy ??
            process.env.HTTP_PROXY);
    }
    else if (protocol === "http:") {
        return (process.env.http_proxy ??
            process.env.HTTP_PROXY ??
            process.env.https_proxy ??
            process.env.HTTPS_PROXY);
    }
    return undefined;
}
export { ConnectionRefusedError, DispatcherError, DownloadError, RequestError, RequestTimeoutError, ResponseStatusCodeError, } from "./errors/request.js";
